---
output:
  pdf_document:
    latex_engine: pdflatex
  html_document:
    df_print: paged
params:
  logo_file: "rapid_cns2_logo.png"
  institution_logo: "institution_logo.png"
header-includes:
- \usepackage{palatino}
- \renewcommand{\familydefault}{\sfdefault} % sans serif
- \fontfamily{ppl}\selectfont
- \usepackage{fancyhdr}
- \pagestyle{fancy}
- \fancyhf{}
- \fancyhead[L]{\includegraphics[height=0.5in]{logos/rapid_cns2_logo.png}}
- \fancyhead[R]{\includegraphics[height=0.5in]{logos/institution_logo.png} \quad \thepage}
- \renewcommand{\headrulewidth}{0.4pt}
geometry: margin=0.2in
always_allow_html: true
---
<style type="text/css">
  body{
  font-size: 14pt;
  text-align: justify;
}
</style>



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(kableExtra)
library(ggplot2)
```

```{r params_setup, echo=FALSE}
# Use the logo parameters from YAML header
# To use custom logos, modify the params section in the YAML header above
# Example: 
# params:
#   logo_file: "rapid_cns2_logo.png"
#   institution_logo: "your_institution_logo.png"
logo_file <- params$logo_file
institution_logo <- params$institution_logo
```

```{r research_use_only, echo=FALSE, results='asis'}
cat("\\begin{center}")
cat("\\vspace{2cm}")
cat("\\textcolor{red}{\\Huge\\textbf{RESEARCH USE ONLY}}")
cat("\\vspace{2cm}")
cat("\\end{center}")
```

```{r check_platform, echo=FALSE}
platform <- "Unknown"
if (seq!="false"){
    platform <- seq
}
if (seq=="F") {
    platform <- "MinION/GridION"
}
if (seq=="P") {
    platform <- "PromethION"
}
if (seq=="P2S") {
    platform <- "P2 Solo"
}
if (seq=="GXB") {
    platform <- "GridION"
}


```

## **rapidCNS^2^ Report**
## **Sample ID:** `r sample` \
## **Patient name:**  `r patient` \

#### This report was generated by **Rapid-CNS^2^ (Nextflow pipeline Rapid-CNS2_nf `r nextflow_ver`)** and is intended for research use only.  \
#### Frozen tumour tissue was prepared and processed using the SQK-LSK114 Ligation Sequencing Kit from Oxford Nanopore Technologies. 160 gene regions from the NPHD panel were targeted using an adaptive sampling approach on the `r platform` platform. Data was analysed by the custom Rapid-CNS^2^ pipeline. Comprehensive analysis of CNS tumours using Rapid-CNS^2^ is a research tool currently under development. It has not been clinically validated in sufficiently large cohorts. Interpretation and implementation of the results in a clinical setting is in the sole responsibility of the treating physician.

## On-target coverage
```{r,echo=FALSE,out.width="75%",warning=FALSE, fig.align='center'}
df <- read.delim(coverage)
df$Reads <- "All"
df$Reads[endsWith(df$chrom,"_region")] <- "On_target"
df$chrom <- gsub("_region","",df$chrom)
df$mean <- as.numeric(df$mean)
df$chrom[df$chrom=="chr1"] <- "chr01"
df$chrom[df$chrom=="chr2"] <- "chr02"
df$chrom[df$chrom=="chr3"] <- "chr03"
df$chrom[df$chrom=="chr4"] <- "chr04"
df$chrom[df$chrom=="chr5"] <- "chr05"
df$chrom[df$chrom=="chr6"] <- "chr06"
df$chrom[df$chrom=="chr7"] <- "chr07"
df$chrom[df$chrom=="chr8"] <- "chr08"
df$chrom[df$chrom=="chr9"] <- "chr09"

df <- df[order(df$chrom),]

df$chrom <- as.character(df$chrom)

df <- df[which(df$chrom!="chrM"),]

df$chrom <- as.factor(df$chrom)
levels(df$chrom) <- unique(df$chrom)
ggplot2::ggplot(data=df, 
  aes(x=chrom,y=mean,color=Reads)) + 
  geom_point(size=3) + 
  theme_minimal() + 
  ylab("Mean coverage") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size = 10), axis.title.y = element_text(size=10)) + 
  ylim(0,max(df$mean))+ scale_color_manual(values = c("#FF5A5F","#0B3954")) + 
  xlab("Chromosome") +theme(text=element_text(family="sans")) + 
  theme(panel.border = element_rect(color = "black", 
                                    fill = NA, 
                                    size = 1))
```
\
\newpage

## Mutations

````{=html}
```{r, echo=FALSE, results='asis', eval=inc_igvreport}
xfun::file_string(igv_report)
```
````

```{r,echo=FALSE, out.width="100%", warning=FALSE, eval=exc_igvreport}
d <- read.csv(mutations)
kable(d, booktabs = TRUE) %>%
  kable_styling(font_size=14, latex_options = c("HOLD_position","scale_down"))
```


\
\newpage

## Copy number profile


```{r, echo=FALSE, out.width="100%",warning=FALSE}
knitr::include_graphics(cnv_plot)
```
\newpage
\
```{r, echo=FALSE}
x <- rf_details
oob <- read.delim(x,header = F, sep = ":")
mc <- read.delim(votes,sep=" ")
```
## Rapid CNS^2^ Methylation classification
#### Methylation-based classification is based on `r as.integer(oob[1,2])` sites. The Brier score of this ad-hoc classifier is `r paste0(round(oob[3,2]*100,digits=1),"%")`. Using this classifier, the sample has been classified as  **`r rownames(mc[which.max(mc$cal_Freq),])`**. This prediction has a confidence score of **`r paste0(round(mc$cal_Freq[which.max(mc$cal_Freq)],digits=1),"%")`**. The calibrated score distribution for the Top 10 entities is given below.

```{r,echo=FALSE,warning=FALSE}
mc$methClass <- rownames(mc)
ggplot(data=mc, aes(x=methClass, y=cal_Freq)) +
  geom_bar(stat="identity", fill="#dc143c") +
  coord_flip() +
  scale_x_discrete(limits=mc[order(mc$cal_Freq,decreasing=T),]$methClass[10:1]) +
  ylim(0,100) +
  ylab("confidence score (%)") + xlab("methylation class") +
  theme_classic() + theme(aspect.ratio=1) +theme(text=element_text(family="sans")) +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 1))

```

\newpage

\newpage
\

## MGMT promoter methylation \
#### Average coverage of MGMT is: **`r cov`**x.\

```{r setup2, echo=FALSE}
show_mgmt <- FALSE
### if methylartist and mgmt exist
if (methylartist_plot == "true" & mgmt == "true")
show_mgmt <- TRUE
```

```{r conditional_block, echo=FALSE, results='asis', eval=show_mgmt}
mgmt_file <- read.csv(opt$mgmt)
avg <- mgmt_file$average
avg <- round(avg, 2)
avg <- paste0(avg,"%")
status <- mgmt_file$status
cat("#### MGMT promoter methylation status is assessed based on a logistic regression based prediction model that considers 137 most predictive CpG sites in the MGMT promoter region. The cutoff is assigned as 25%.")
cat("\n\n#### Average methylation = ")
cat(avg)
cat("\n\n#### Predicted MGMT promoter status = ")
cat(status)
```

```{r setup3, echo=FALSE}
mgmt_too_low <- FALSE
## if methylartist and mgmt exist
if (!file.exists(paste0(opt$outdir, "/*.chr10_129466536_129467536.m.ms1.smw20.locus.meth.png")) 
    & !file.exists(paste0(opt$outdir, "/", opt$sample, "_mgmt_status.csv")))
mgmt_too_low <- TRUE
```
```{r, echo=FALSE, warning=FALSE, results='asis', eval=mgmt_too_low}
cat("#### MGMT promoter methylation status was not assessed in this sample due to sequence coverage at the MGMT promter region being too low.")
```

## MGMT Promoter methylation plot:
```{r, echo=FALSE, out.width="100%",warning=FALSE, eval=show_mgmt}
knitr::include_graphics(opt$methylartist)
```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=mgmt_too_low}
cat("#### An MGMT promoter plot was not produced in this sample due to sequence coverage at the MGMT site being too low.")
```


#### The Rapid CNS2 classification and report generation followed the protocol designed and published by Patel _et_ _al._ (2022).

#### Patel, A., Dogan, H., Payne, A., Krause, E., Sievers, P., Schoebe, N., Schrimpf, D., Blume, C., Stichel, D., Holmes, N., Euskirchen, P., Hench, J., Frank, S., Rosenstiel-Goidts, V., Ratliff, M., Etminan, N., Unterberg, A., Dieterich, C., Herold-Mende, C., Pfister, S. M., … Sahm, F. (2022). Rapid-CNS^2^: rapid comprehensive adaptive nanopore-sequencing of CNS tumors, a proof-of-concept study. _Acta_ _neuropathologica_, _143_(5), 609–612. https://doi.org/10.1007/s00401-022-02415-6 \

#### **The information contained with this report is intended for research use only and is not for use in diagnostic procedures.**

#### Email: a.patel@kitz-heidelberg.de  \
#### Date: `r format(Sys.time(), '%d %B, %Y')`

