---
output:
  pdf_document:
    latex_engine: pdflatex
    includes:
      header: |
        \usepackage{graphicx}
        \usepackage{fancyhdr}
        \addtolength{\headheight}{3.0cm}
        \pagestyle{fancyplain}
        \graphicspath{{./logos/}}
        \fancyhead[R]{\includegraphics[height=3cm]{rapid_cns2_logo.png}}
        \fancyhead[L]{\includegraphics[height=3cm]{ukhd_logo.png}}
        \renewcommand{\headrulewidth}{0pt}
        \usepackage{helvet}
        \renewcommand{\familydefault}{\sfdefault}
        \fontfamily{phv}\selectfont
        \usepackage{booktabs}
        \usepackage{longtable}
        \usepackage{array}
        \usepackage{multirow}
        \usepackage{wrapfig}
        \usepackage{float}
        \usepackage{colortbl}
        \usepackage{pdflscape}
        \usepackage{tabu}
        \geometry{margin=0.5in, top=1in, bottom=1in}

params:
  software_ver: null

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(kableExtra)
library(ggplot2)

# Copy logos from container to logos directory for PDF rendering
if (!dir.exists("logos")) {
  dir.create("logos")
}
if (file.exists("/opt/logos/rapid_cns2_logo.png")) {
  file.copy("/opt/logos/rapid_cns2_logo.png", "logos/rapid_cns2_logo.png", overwrite = TRUE)
}
if (file.exists("/opt/logos/ukhd_logo.png")) {
  file.copy("/opt/logos/ukhd_logo.png", "logos/ukhd_logo.png", overwrite = TRUE)
}
```



```{r check_platform, echo=FALSE}
platform <- "Unknown"
if (seq!="false"){
    platform <- seq
}
if (seq=="F") {
    platform <- "MinION/GridION"
}
if (seq=="P") {
    platform <- "PromethION"
}
if (seq=="P2S") {
    platform <- "P2 Solo"
}
if (seq=="P2i") {
    platform <- "P2 Integrated"
}
if (seq=="GXB") {
    platform <- "GridION"
}
```

# **Rapid-CNS^2^ Report**

**Sample ID:** `r sample`  
**Patient:** `r patient`

This report was generated by **Rapid-CNS^2^ (Nextflow pipeline Rapid-CNS2_nf `r software_ver`)** and is intended for research use only.

Frozen tumour tissue was prepared and processed using the SQK-LSK114 Ligation Sequencing Kit from Oxford Nanopore Technologies. 160 gene regions from the NPHD panel were targeted using an adaptive sampling approach on the `r platform` platform. Data was analysed by the custom Rapid-CNS^2^ pipeline. Comprehensive analysis of CNS tumours using Rapid-CNS^2^ is a research tool currently under development. It has not been clinically validated in sufficiently large cohorts. Interpretation and implementation of the results in a clinical setting is in the sole responsibility of the treating physician.

\newpage

# On-target coverage

```{r,echo=FALSE,warning=FALSE, fig.align='center'}
df <- read.delim(coverage)
df$Reads <- "All"
df$Reads[endsWith(df$chrom,"_region")] <- "On_target"
df$chrom <- gsub("_region","",df$chrom)
df$mean <- as.numeric(df$mean)
df$chrom[df$chrom=="chr1"] <- "chr01"
df$chrom[df$chrom=="chr2"] <- "chr02"
df$chrom[df$chrom=="chr3"] <- "chr03"
df$chrom[df$chrom=="chr4"] <- "chr04"
df$chrom[df$chrom=="chr5"] <- "chr05"
df$chrom[df$chrom=="chr6"] <- "chr06"
df$chrom[df$chrom=="chr7"] <- "chr07"
df$chrom[df$chrom=="chr8"] <- "chr08"
df$chrom[df$chrom=="chr9"] <- "chr09"

df <- df[order(df$chrom),]

ggplot(df, aes(x=chrom, y=mean, fill=Reads)) +
  geom_bar(stat="identity", position="dodge") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ylab("Mean coverage") +
  xlab("Chromosome") +
  scale_fill_manual(values=c("All"="lightblue", "On_target"="darkblue"))
```

\newpage

# Mutations

```{r, echo=FALSE}
if (file.exists(mutations)) {
  mutations_df <- read.csv(mutations)
  if (nrow(mutations_df) > 0) {
    kable(mutations_df, booktabs = TRUE) %>%
      kable_styling(font_size=10, latex_options = c("HOLD_position","scale_down"))
  } else {
    cat("No mutations found.")
  }
} else {
  cat("Mutations file not found.")
}
```

\newpage

# Copy number profile

```{r, echo=FALSE, warning=FALSE}
knitr::include_graphics(cnv_plot)
```

\newpage

# Rapid CNS^2^ Methylation classification

```{r, echo=FALSE}
x <- rf_details
oob <- read.delim(x,header = F, sep = ":")
mc <- read.delim(votes,sep=" ")
```

Methylation-based classification is based on `r as.integer(oob[1,2])` sites. The Brier score of this ad-hoc classifier is `r paste0(round(as.numeric(oob[3,2])*100,digits=1),"%")`. Using this classifier, the sample has been classified as **`r rownames(mc[which.max(mc$Recalibrated_Frequency),])`**. This prediction has a confidence score of **`r paste0(round(as.numeric(mc$cal_Freq[which.max(mc$cal_Freq)]),digits=1),"%")`**. The calibrated score distribution for the Top 10 entities is given below.

```{r,echo=FALSE,warning=FALSE}
mc$methClass <- rownames(mc)
ggplot(data=mc, aes(x=methClass, y=Recalibrated_Frequency)) +
  geom_bar(stat="identity", fill="#dc143c") +
  coord_flip() +
  scale_x_discrete(limits=mc[order(mc$Recalibrated_Frequency,decreasing=T),]$methClass[10:1]) +
  ylim(0,100) +
  ylab("confidence score (%)") + xlab("methylation class") +
  theme_classic() + theme(aspect.ratio=1) +theme(text=element_text(family="sans")) +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 1))
```

\newpage

# MGMT promoter methylation

Average coverage of MGMT is: **`r cov`**x.

```{r setup2, echo=FALSE}
show_mgmt <- FALSE
### if methylartist and mgmt exist
if (methylartist_status == "true" & mgmt_status == "true")
show_mgmt <- TRUE
```

```{r conditional_block, echo=FALSE, results='asis', eval=show_mgmt}
mgmt_file <- read.csv(mgmt)
avg <- mgmt_file$average
avg <- round(as.numeric(avg), 2)
avg <- paste0(avg,"%")
status <- mgmt_file$status
cat("MGMT promoter methylation status is assessed based on a logistic regression based prediction model that considers 137 most predictive CpG sites in the MGMT promoter region. The cutoff is assigned as 25%")
cat("\n\nAverage methylation = ")
cat(avg)
cat("\n\nPredicted MGMT promoter status = ")
cat(status)
```

```{r setup3, echo=FALSE}
mgmt_too_low <- FALSE
## if methylartist and mgmt exist
if (!file.exists(methylartist_plot) 
    & !file.exists(mgmt))
mgmt_too_low <- TRUE
```
```{r, echo=FALSE, warning=FALSE, results='asis', eval=mgmt_too_low}
cat("MGMT promoter methylation status was not assessed in this sample due to sequence coverage at the MGMT promter region being too low.")
```

## MGMT Promoter methylation plot:

```{r, echo=FALSE, warning=FALSE, eval=show_mgmt}
knitr::include_graphics(methylartist_plot)
```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=mgmt_too_low}
cat("An MGMT promoter plot was not produced in this sample due to sequence coverage at the MGMT site being too low.")
```

\newpage

\small

The Rapid CNS2 classification and report generation followed the protocol designed and published by Patel \textit{et al.} (2022) and Patel \textit{et al.} (2025).

Patel, A., Dogan, H., Payne, A., Krause, E., Sievers, P., Schoebe, N., Schrimpf, D., Blume, C., Stichel, D., Holmes, N., Euskirchen, P., Hench, J., Frank, S., Rosenstiel-Goidts, V., Ratliff, M., Etminan, N., Unterberg, A., Dieterich, C., Herold-Mende, C., Pfister, S. M., … Sahm, F. (2022). Rapid-CNS^2^: rapid comprehensive adaptive nanopore-sequencing of CNS tumors, a proof-of-concept study. \textit{Acta neuropathologica}, \textit{143}(5), 609–612. https://doi.org/10.1007/s00401-022-02415-6

Patel, A., Göbel, K., Ille, S. et al. Prospective, multicenter validation of a platform for rapid molecular profiling of central nervous system tumors. \textit{Nature Medicine} 31, 1567–1577 (2025). https://doi.org/10.1038/s41591-025-03562-5

\textbf{The information contained with this report is intended for research use only and is not for use in diagnostic procedures.}

For any questions, please contact:
\textbf{Areeba Patel, PhD}

Department of Neuropathology, German Cancer Research Center

\href{mailto:a.patel@dkfz.de}{a.patel@dkfz.de}

\textbf{Prof Dr. Dr. med Felix Sahm}

Head of Molecular Neuropathology, University Hospital Heidelberg

\href{mailto:felix.sahm@med.uni-heidelberg.de}{felix.sahm@med.uni-heidelberg.de}

Report generated on `r format(Sys.time(), '%d %B, %Y at %H:%M')` 