---
output:
  pdf_document:
    latex_engine: pdflatex
  html_document:
    df_print: kable
header-includes:
- \usepackage{fancyhdr}
- \addtolength{\headheight}{3.0cm} % make more space for the header
- \pagestyle{fancyplain} % use fancy for all pages except chapter start
- \fancyhead[R]{\includegraphics[height=3cm]{/b06x-isilon/b06x-m/mnp_nanopore/git_repos/Rapid-CNS2/scr/171030Logo_UKHD_engl_positiv_RGB.png}}
- \renewcommand{\headrulewidth}{0pt}
- \usepackage{palatino}
- \renewcommand{\familydefault}{\sfdefault} % sans serif
- \fontfamily{ppl}\selectfont
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
geometry: "left=1.5cm,right=1.5cm,top=2cm,bottom=2cm"
always_allow_html: true
---
<style>
  body{
  font-size: 16pt;
  text-align: justify;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(kableExtra)
library(ggplot2)
```

\thispagestyle{empty}
```{r, echo=FALSE,out.width="100%",warning=FALSE}
knitr::include_graphics("/b06x-isilon/b06x-m/mnp_nanopore/paper/report/Screenshot 2022-03-17 at 6.28.43 PM.png")
```

\
**Sample ID:** `r sample` \
**Patient name:**  `r patient`

\
This report was generated by **Rapid-CNS^2^** and is intended for research use only. Frozen tumour tissue was prepared and processed using the SQK-LSK109 Ligation Sequencing Kit from Oxford Nanopore Technologies. 160 gene regions from the NPHD panel and 10,000 CpG sites (plus 10kb flanking region) were targeted using an adaptive sampling approach on the GridION. Data was analysed by the custom Rapid-CNS^2^ pipeline. Comprehensive analysis of CNS tumours using Rapid-CNS^2^ is a research tool currently under development. It has not been clinically validated in sufficiently large cohorts. Interpretation and implementation of the results in a clinical setting is in the sole responsibility of the treating physician.
\

**On-target coverage**

```{r,echo=FALSE,out.width="85%",warning=FALSE,fig.width=12,fig.height=7}
df <- read.delim(coverage)
df$Reads <- "All"
df$Reads[endsWith(df$chrom,"_region")] <- "On_target"
df$chrom <- gsub("_region","",df$chrom)
df$mean <- as.numeric(df$mean)
df$chrom <- as.factor(df$chrom)
levels(df$chrom) <- unique(df$chrom)
ggplot2::ggplot(data=df, aes(x=chrom,y=mean,color=Reads)) + geom_point(size=6) + theme_minimal(base_size=14) + ylab("Mean coverage") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size = 14), axis.title.y = element_text(size=14)) + ylim(0,70)+ scale_color_manual(values = c("#FF5A5F","#0B3954")) + xlab("Chromosome") +theme(text=element_text(family="sans"))
```

\newpage

**Mutations**

```{r,echo=FALSE,warning=FALSE,results='asis'}
d <- read.csv(mutations)

knitr::kable(d, booktabs=T,format = "latex",longtable = TRUE) %>% 
  kable_styling(position="center", font_size=7,latex_options = c("HOLD_position","repeat_header"))
```

\newpage

**Copy number profile**


```{r, echo=FALSE, out.width="100%",warning=FALSE}
knitr::include_graphics(cnv_plot)
```
\newpage

```{r, echo=FALSE}
x <- rf_details
oob <- read.delim(x,header = F, sep = ":")
mc <- read.delim(votes,sep=" ")
```
**Methylation classification**

Methylation-based classification is based on **`r as.integer(oob[1,2])`** sites. The Brier score of this ad-hoc classifier is **`r paste0(round(oob[3,2]*100,digits=1),"%")`**. Using this classifier, the sample has been classified as  **`r rownames(mc[which.max(mc$cal_Freq),])`**. This prediction has a confidence score of **`r paste0(round(mc$cal_Freq[which.max(mc$cal_Freq)],digits=1),"%")`**. The calibrated score distribution for the Top 10 entities is given below.

```{r,echo=FALSE,warning=FALSE}
mc$methClass <- rownames(mc)
ggplot(data=mc, aes(x=methClass, y=cal_Freq)) +
  geom_bar(stat="identity", fill="#dc143c") +
  coord_flip() +
  scale_x_discrete(limits=mc[order(mc$cal_Freq,decreasing=T),]$methClass[10:1]) +
  ylim(0,100) +
  ylab("confidence score (%)") + xlab("methylation class") +
  theme_classic() + theme(aspect.ratio=1) +theme(text=element_text(family="sans"))

```
```{r,echo=FALSE,warning=FALSE}
mgmt <- read.csv(mgmt)
avg <- mgmt$average
avg <- paste0(avg,"%")
status <- mgmt$status
```

**MGMT promoter methylation** \
MGMT promoter methylation status is assessed based on a logistic regression based prediction model that considers 137 most predictive CpG sites in the MGMT promoter region. The cutoff is assigned as 25%. \
\
Average methylation = **`r avg`** \
Predicted MGMT promoter status= **`r status`** \
\
\
\
\
\
\
Author: Areeba Patel, Department of Neuropathology, University Hospital Heidelberg \
Email: areeba.patel@med.uni-heidelberg.de \
Date: `r format(Sys.time(), '%d %B, %Y')`