---
output:
  pdf_document:
    latex_engine: pdflatex
  html_document:
    df_print: kable
header-includes:
- \usepackage{fancyhdr}
- \addtolength{\headheight}{1.75cm} % make more space for the header
- \pagestyle{fancyplain} % use fancy for all pages except chapter start
- \fancyhead[L]{\includegraphics[height=2cm]{/hospital_en.jpeg}}
- \fancyhead[R]{Department of Neuropathology\\Prof. Dr. med. A. von Deimling\\INF 224, 69120 Heidelberg, Germany}
- \renewcommand{\headrulewidth}{0.5pt}
- \usepackage{palatino}
- \renewcommand{\familydefault}{\sfdefault} % sans serif
- \fontfamily{ppl}\selectfont
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{graphicx}
- \setlength\parindent{24pt}
geometry: "left=1.5cm,right=1.5cm,top=2cm,bottom=2cm"
always_allow_html: true
---
<style>
  body{
  font-size: 12pt;
  text-align: justify;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(kableExtra)
library(ggplot2)
```

\thispagestyle{empty}
\
```{r, echo=FALSE,out.width="60%",warning=FALSE, fig.align="center"}
knitr::include_graphics("/header_logo.png")
```
\centering{\section{\LARGE Neuropathological Report generated by Rapid-CNS$^2$}}

\raggedright
\
\
\subsection{General Information}
\begin{table}[H]
\begin{tabular}{lll}
\multicolumn{3}{l}{\textit{Patient Information}} \\
    & Sample ID                  & \textbf{`r sample`}    \\
    & Tissue                     & \textbf{`r tissue`} \\	    
    & Library                    & \textbf{`r gsub("_","\\\\_",library)`}   \\
    & Sequencing date            & \textbf{`r seqdate`}   \\
    &                            &               \\  
\multicolumn{3}{l}{\textit{Library Information}} \\
    & Sequencing preparation kit    & \textbf{`r gsub("-","\\\\-",prepkit)`}   \\
    & Panel                      & \textbf{`r gsub("_","\\\\_",panel)`}   \\
    & Number of genes            & \textbf{`r num_genes`}   \\
    & Reference genome           & \textbf{`r gsub("_","\\\\_",refgen)`}          \\
    & Device                     & \textbf{`r device`}       \\
\end{tabular}
\end{table}


\
\subsection{Disclaimer} \
This report was automatically generated by Rapid-CNS^2^ and is currently intended for research use only. The sample tissue was prepared and processed using the aforementioned library sequencing kit from Oxford Nanopore Technologies. Data was analysed by the custom Rapid-CNS^2^ pipeline, an overview of the used tools can be found in the attachments of this report. Additionally, the genes included in the used panel can be found there. \
Comprehensive analysis of CNS tumours using Rapid-CNS^2^ is a research tool currently under development. It has not been clinically validated in sufficiently large cohorts. Interpretation and implementation of the results in a clinical setting is in the sole responsibility of the treating physician.
\
\
Author: Kirsten GÃ¶bel, Department of Neuropathology, University Hospital Heidelberg \
Email: kirsten.goebel@med.uni-heidelberg.de \
Based on the work of Areeba Patel (a.patel@dkfz-heidelberg.de)
\
\
\
\textbf{Analysis Date: `r format(Sys.time(), '%d %B, %Y')`} \


\newpage

\subsection{On-target coverage}
```{r,echo=FALSE,out.width="85%",warning=FALSE,fig.width=12,fig.height=7}
df <- read.delim(coverage)
df$Reads <- "All"
df$Reads[endsWith(df$chrom,"_region")] <- "On_target"
df$chrom <- gsub("_region","",df$chrom)
df <- subset(df, chrom %in% c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr11", "chr10", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY"))
df$mean <- as.numeric(df$mean)
df$chrom <- as.factor(df$chrom)
levels(df$chrom) <- unique(df$chrom)
ggplot2::ggplot(data=df, aes(x=chrom,y=mean,color=Reads)) + geom_point(size=6) + theme_minimal(base_size=14) + ylab("Mean coverage") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size = 14), axis.title.y = element_text(size=14)) + ylim(0,70)+ scale_color_manual(values = c("#FF5A5F","#0B3954")) + xlab("Chromosome") +theme(text=element_text(family="sans"))

tmp <- df[df$Reads == "On_target",]
paste("Mean on-target coverage:", toString(mean(tmp$mean))) 
```

\newpage

\subsection{Copy number profile}


```{r, echo=FALSE, out.width="100%",warning=FALSE}
knitr::include_graphics(cnv_plot)
```
\newpage

```{r, echo=FALSE}
x <- rf_details
oob <- read.delim(x,header = F, sep = ":")
mc <- read.delim(votes,sep=" ")
```
\subsection{Methylation classification}

Methylation-based classification is based on **`r as.integer(oob[1,2])`** sites. The Brier score of this ad-hoc classifier is **`r paste0(round(oob[3,2]*100,digits=1),"%")`**. Using this classifier, the sample has been classified as  **`r rownames(mc[which.max(mc$cal_Freq),])`**. This prediction has a confidence score of **`r paste0(round(mc$cal_Freq[which.max(mc$cal_Freq)],digits=1),"%")`**. The calibrated score distribution for the Top 10 entities is given below.

# ```{r,echo=FALSE,warning=FALSE}
# mc$methClass <- rownames(mc)
# ggplot(data=mc, aes(x=methClass, y=cal_Freq)) +
#  geom_bar(stat="identity", fill="#dc143c") +
#  coord_flip() +
#  scale_x_discrete(limits=mc[order(mc$cal_Freq,decreasing=T),]$methClass[10:1]) +
#  ylim(0,100) +
#  ylab("confidence score (%)") + xlab("methylation class") +
#  theme_classic() + theme(aspect.ratio=1) +theme(text=element_text(family="sans"))
# ```
```{r, echo=FALSE, out.width="100%",warning=FALSE}
knitr::include_graphics(classification_plot)
```

```{r,echo=FALSE,warning=FALSE}
if (file.exists(mgmt)) {
  mgmt <- read.csv(mgmt)
  avg <- mgmt$average
  avg <- paste0(avg,"%")
  status <- mgmt$status
  } else {
  avg <- 0
  status="unknown"
  }
```
\subsection{MGMT promoter methylation}
MGMT promoter methylation status is assessed based on a logistic regression based prediction model that considers 137 most predictive CpG sites in the MGMT promoter region. The cutoff is assigned as 25%. \
\
Average methylation = **`r avg`** \
Predicted MGMT promoter status= **`r status`** \

\newpage

\subsection{Variants}
The variants in this report are in-silico filtered down to the genes included in the aforementioned panel. 
\subsection{Special Positions}
```{r,echo=FALSE,warning=FALSE,results='asis'}

d <- read.xlsx(specpos, check.names=FALSE)
d$ALT <- gsub("&lt;", "", d$ALT)
d$ALT <- gsub("&gt;", "", d$ALT)

knitr::kable(d, booktabs=T,format = "latex",longtable = TRUE) %>%
  kable_styling(position="center", font_size=8,latex_options = c("HOLD_position","repeat_header"))

```
\newpage
\subsubsection{SNVs}

```{r,echo=FALSE,warning=FALSE,results='asis'}
d <- read.csv(snv)

knitr::kable(d, booktabs=T,format = "latex",longtable = TRUE) %>%
  kable_styling(position="center", font_size=8,latex_options = c("HOLD_position","repeat_header"))

```

\subsubsection{InDels}
\pagestyle{empty}
```{r,echo=FALSE,warning=FALSE,results='asis'}
d <- read.csv(indel)

landscape(knitr::kable(d, booktabs=T,format = "latex",longtable = TRUE) %>%
  kable_styling(position="center", font_size=8,latex_options = c("HOLD_position","repeat_header", "scale_down")))

```
Allele Frequency of the Population <= 0.01, based on 1000 genomes.
\pagestyle{fancy}
\subsubsection{Fusions}

```{r,echo=FALSE,warning=FALSE,results='asis'}

if (file.exists(fusions)) {
knitr::kable(d, booktabs=T,format = "latex",longtable = TRUE) %>%
  kable_styling(position="center", font_size=8,latex_options = c("HOLD_position","repeat_header")) 
  } else {
    "fusion calling not performed"
  }

```
\newpage


\section{Attachments}
\subsection{Quality Information}
soon to come
\subsection{Tools and Versions}
\begin{table}[H]
\centering
\begin{tabular}{@{}lll@{}}
\toprule
\textbf{tool}              & \textbf{version} & \textbf{citation}                             \\ \midrule
Guppy Basecalling Software & 6.4.8+31becc9    & Oxford Nanopore Technologies plc.             \\
pycoQC & pycoQC v2.5.2    & https://doi.org/10.21105/joss.01236             \\
samtools                   & 1.16.1           & https://doi.org/10.1093/gigascience/giab008   \\
liftOver                   & -                & UCSC                                          \\
bedtools                   & v2.30.0          & https://doi.org/10.1093/bioinformatics/btq033 \\
cnvpytor                   & 1.2.1            & https://doi.org/10.1093/gigascience/giab074   \\
mosdepth                   & v0.3.3           & https://doi.org/10.1093/bioinformatics/btx699 \\
longshot                   & v0.4.5           & https://doi.org/10.1038/s41467-019-12493-y    \\
P.E.P.P.E.R.               & r0.8             & https://doi.org/10.1038/s41592-021-01299-w    \\
svim                       & v2.0.0           & https://doi.org/10.1093/bioinformatics/btz041 \\ \bottomrule
\end{tabular}
\end{table}

\subsection{Genes of Panel}
